File: src/main/java/util/PowerSet.java
package util;

import java.util.Iterator;
import java.util.List;
import java.util.stream.Collectors;

public class PowerSet<T> implements Iterator<List<T>> {
    boolean leftMode = true;
    List<T> memory = null;
    PowerSet<T> delegate;
    T pivot;

    public PowerSet(Iterator<T> storage) {
        if (storage.hasNext()) {
            this.pivot = storage.next();
            this.delegate = new PowerSet<>(storage);
        } else {
            this.delegate = null;
            this.pivot = null;
        }
    }

    public static void main(String[] args) {
        var input = List.of(5,5,3,2,1);
        var ps = new PowerSet<>(input.iterator());

        while (ps.hasNext()) {
            System.out.println(ps.next().stream().map(Object::toString).collect(Collectors.joining(",")));
        }
    }

    public boolean hasNext() {
        if (delegate == null) {
            boolean hasNext = leftMode;
            leftMode = false;
            return hasNext;
        }
        if (leftMode) {
            return delegate.hasNext();
        } else {
            return true;
        }
    }

    public List<T> next() {
        if (delegate == null) {
            return java.util.Collections.emptyList();
        }
        if (leftMode) {
            memory = delegate.next();
            leftMode = false;
            return memory;
        } else {
            leftMode = true;
            var list = new java.util.ArrayList<T>();
            list.add(pivot);
            list.addAll(memory);
            return list;
        }
    }

}

File: src/main/kotlin/PowerSet.kt
import java.util.*

fun main() {
    listOf(5, 5, 3, 2, 1).apply {
        PowerSet(iterator())
            .apply { while (hasNext()) println(next().joinToString(",") { it.toString() }) }
    }
}

class PowerSet<T>(storage: Iterator<T>) : Iterator<List<T>> {

    private var leftMode = true
    private var memory = Optional.empty<List<T>>()
    private var delegate = Optional.empty<PowerSet<T>>()
    private var pivot = Optional.empty<T>()

    override fun hasNext(): Boolean {
        return when {
            delegate.isEmpty -> {
                val hasNext = leftMode
                leftMode = false
                hasNext
            }
            leftMode -> delegate.get().hasNext()
            else -> true
        }
    }

    override fun next(): List<T> {
        return when {
            delegate.isEmpty -> emptyList()
            else -> when {
                leftMode -> Optional.of(delegate.get().next())
                    .also { memory = it }
                    .also { leftMode = false }.get()
                else -> ArrayList<T>()
                    .apply { add(pivot.get()) }
                    .apply { addAll(memory.get()) }
                    .also { leftMode = true }
            }
        }
    }

    init {
        when {
            storage.hasNext() -> {
                pivot = Optional.of(storage.next()!!)
                delegate = Optional.of(PowerSet(storage))
            }
            else -> {
                delegate = Optional.empty()
                pivot = Optional.empty()
            }
        }
    }
}